pkg/endpoint/bpf.go

```diff
@@ -1283,14 +1283,16 @@ func (e *Endpoint) syncPolicyMap() error {
 	}
 
 	// Diffs between the maps are expected here, so do not bother collecting them
-	_, _, err = e.syncDesiredPolicyMapWith(e.realizedPolicy.PolicyMapState, false)
+	_, _, err = e.syncPolicyMapsWith(e.realizedPolicy.PolicyMapState, false)
 	return err
 }
 
-// syncDesiredPolicyMapWith updates the bpf policy map state based on the
+// syncPolicyMapsWith updates the bpf policy map state based on the
 // difference between the given 'realized' and desired policy state without
 // dumping the bpf policy map.
-func (e *Endpoint) syncDesiredPolicyMapWith(realized policy.MapState, withDiffs bool) (diffCount int, diffs []policy.MapChange, err error) {
+// Changes are synced to endpoint's realized policy mapstate, 'realized' is
+// not modified.
```

Comment:
I'm not sure I quite follow the meaning in this last sentence.

It looks to me like we go through the desired policy map, and apply all desired changes that the daemon thinks have not yet been realized. So it's more that the _bpf_ policymaps are synced to the _desired_ state based on the daemon's understanding of the realized state.

Reply:
The whole terminology here of "sync desired policy map with ( realized )" just seems kinda confusing to me. The goal here is to sync the BPF policy maps with the desired state, right? But we just make it more efficient by using the current realized state view.

Reply:
(thinking aloud if there's another less confusing naming convention here..)

Reply:
I think that part of my confusion above is that "sync <x> with <y>" has specific connotations about the way that the information is flowing, basically that <x> is the thing that changes and <y> is the source of truth. But right now, the thing that actually changes (bpf) is not explicitly mentioned, and it's implied that the desired policy changes here... which doesn't seem to be the case.

Reply:
Looking a little further, with the latest version of this PR, this statement seems to be incorrect, since `addPolicyKey()` and `deletePolicyKey()` are still modifying the `realizedPolicy`: `'realized' is not modified.`

Reply:
This function can be called with a temporary 'realized' map, dumped from the datapath, or with the endpoint's realized policy map. Either way diffs between this passed in map and endpoint's desired policy map are applied to the bpf policy map *and* the endpoint's realized policy map, not the passed in 'realized' map. For the call where 'realized' is the endpoint's realized policy map this does not mean much, but in the case where 'realized' is a temporary map dumped from the datapath there is no point updating it after computing the diff (as the temporary map is thrown away right after), and this way the endpoint's own realized policy map becomes (more) synced with the datapath bpf map.

Reply:
Ah, okay - I was wondering why this 'realized' map was explicitly handed in, as I assumed that this `realized` map was exactly the same as `e.realizedPolicy`, but that's not the case. This stuff is pretty subtle :thinking: